drivers/include/periph/gpio_ll.h

```diff
@@ -0,0 +1,417 @@
+ * @brief       Next Generation Peripheral GPIO Interface
+ *
+ * @warning     This API is not stable yet and intended for internal use only
+ *              as of now.
+ *
+ * # Design Goals
+ *
+ * This API aims to provide low level access to GPIOs with as little
+ * abstraction and overhead in place as possible for the hot code paths, while
+ * providing a relatively high-level and feature complete API for the
+ * configuration of GPIO pins. The former is to enable sophisticated use cases
+ * such at bit-banging parallel protocols, bit-banging at high data rates,
+ * bit-banging with strict timing requirements, or any combination of these.
+ * The latter is to expose as much of the features the (arguably) most
+ * important peripheral of the MCU as possible.
+ *
+ * It is possible to implement the high level pin-based GPIO API of RIOT, @ref
+ * drivers_periph_gpio, on top of this API. It is expected that for many use
+ * cases the high level API will still remain the API of choice, since it is
+ * more concise an easier to use.
+ *
+ * @{
+ * @file
+ * @brief       Next Generation Peripheral GPIO Interface
+ *
+ * @author      Gunar Schorcht <gunar@schorcht.net>
+ * @author      Marian Buschsieweke <marian.buschsieweke@ovgu.de>
+ *
+ * @warning     This API is not stable yet and intended for internal use only as
+ *              of now.
+ */
+
+#ifndef PERIPH_GPIO_LL_H
+#define PERIPH_GPIO_LL_H
+
+#include <inttypes.h>
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "architecture.h"
+#include "periph_cpu.h"
+#include "periph/gpio.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#if !defined(HAVE_GPIO_PORT_T) || defined(DOXYGEN)
+/**
+ * @brief   GPIO port type
+ */
+typedef uintptr_t gpio_port_t;
+#endif
+
+#if !defined(GPIO_PORT_UNDEF) || defined(DOXYGEN)
+/**
+ * @brief   Magic "undefined GPIO port" value
+ */
+#define GPIO_PORT_UNDEF         UINTPTR_MAX
+#endif
+
+#ifdef DOXYGEN
+/**
+ * @brief   Get the @ref gpio_port_t value of the port identified by @p num
+ *
+ * @note    If @p num is a compile time constant, this is guaranteed to be
+ *          suitable for a constant initializer.
+ *
+ * Typically this will be something like `(GPIO_BASE_ADDR + num * sizeof(struct
+ * vendor_gpio_reg))`
+ */
+#define GPIO_PORT(num)  implementation_specific
+#endif
+
+#ifdef DOXYGEN
+/**
+ * @brief   Get the number of the GPIO port belonging to the given @ref
+ *          gpio_port_t value
+ *
+ * @note    If @p port is a compile time constant, this is guaranteed to be
+ *          suitable for a constant initializer.
+ *
+ * @pre     @p port is the return value of @ref GPIO_PORT
+ *
+ * For every supported port number *n* the following `assert()` must not blow
+ * up:
+ *
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~{.c}
+ * assert(n == GPIO_PORT_NUM(GPIO_PORT(n)));
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ */
+#define GPIO_PORT_NUM(port) implementation_specific
+#endif
+
```

Comment:
Should we have an additional define `GPIO_PORT_NUMOF` here so that `assert(GPIO_PORT(n) < GPIO_PORT_NUMOF)` can be used?

Reply:
That would be useful to have, but it would also be effort to add this and maintain it. Some platforms will be able to extract this from the vendor header files, but not all.

Reply:
So you're leaving it optional for implementation?

Reply:
Sorry for the long stall. I think that an `GPIO_PORT_IS_VALID()` is more useful, as some stripped down MCU versions drop a GPIO port with functionality deemed unneeded for a lower end version, but this port might not be the last port. E.g. the ATmega32U4 does not have a GPIO Port A, but has Ports B-F (with some ports not having a full number of 8 pins).

Reply:
Do you plan to add something like `GPIO_PORT_IS_VALID` to the API in this PR?